### PR TITLE
Simplify EqHistColorMapper calculations

### PR DESCRIPTION
**Pull request by ianthomas23**
This PR simplifies the maths in the `EqHistColorMapper` class.

The `scan` function takes an array of numbers and calculates a fine-grained histogram of them. The cumulative sum of this gives the CDF (Cumulative Distribution Function). The `EqHist` idea is to split the CDF into a number of equal-sized bins, one per color in the palette, to use the colors as evenly as possible regardless of distribution of the underlying data.

Previously there was an iterative process to find the color bins. I have replaced this with a simple interpolation which is easier to understand and faster.

I don't really know what to do about demonstrating its correctness. I am inclined to invoke the mathematicians "by inspection" excuse.

### CODE REVIEW

**Review by jbednar — COMMENTED** [review]
Looks too good to be true! We definitely struggled to get this implemented initially, and if this is actually valid, fabulous! Looking at the test output, the only visible difference I can see is on .../test/baselines/linux/ColorBar_annotation__should_support_EqHistColorMapper.png , and I can't tell which version (before or after this change) is better. For the rest, the visual images look identical, but the numerical values in the color bar differ, which seems fine (as they must not differ much).

**Comment by bryevdv** [comment]
> I don't really know what to do about demonstrating its correctness. I am inclined to invoke the mathematicians "by inspection" excuse.

For *correctness* I would agree in this case, but for maintainability, it would be nice to have aa few basic numerical tests just to document the behavior and ensure that is does not change unintentionally.

**Comment by mattpap** [comment]
> For correctness I would agree in this case, but for maintainability, it would be nice to have aa few basic numerical comparison tests just to document the behavior and ensure that is does not change unintentionally in the future.

This indeed needs to be sealed in with some numerical unit tests.

File: bokehjs/src/lib/models/mappers/eqhist_color_mapper.ts [hunk 1]
```diff
@@ -43,45 +42,14 @@ export class EqHistColorMapper extends ScanningColorMapper {
...
-    } else {
-      binning = binning.slice(binning.length - n - 1)
-      if (finite_bins != n)
-        logger.warn("EqHistColorMapper warning: Histogram equalization did not converge.")
-    }
-
-    // XXX: should this be guaranteed by the above algorithm?
+    // Ensure bin limits exactly match low and high values
     binning[0] = low
     binning[binning.length-1] = high
```
[the remark this answers is not in the document]
Comment (ianthomas23) [reply bokehjs/src/lib/models/mappers/eqhist_color_mapper.ts:54]: It does need to be manually enforced. The CDF binning uses `eq_bin_centers` of the input data histogram, code which I have not altered. The returned `binning[0]` and `binning[-1]` therefore do not include the data min and max limits. Without these two lines the chosen colors are the same, but the colorbar has the wrong min and max limits.

The algorithm could be rewritten to produce these boundary values but I would not want to do that without also making similar changes to datashader, upon which this code was/is based, and that is much wider scope that I am attempting here.

**Comment by ianthomas23** [comment]
> This indeed needs to be sealed in with some numerical unit tests.

I can add bokehjs unit tests that use `v_compute()`, like the other `ColorMapper` classes. But to add numerical unit tests I would need to call `scan()` which is protected. Is there a recommended way of doing this within the unit testing framework?

**Comment by mattpap** [comment]
>  Is there a recommended way of doing this within the unit testing framework?

Normally we would just unprotect a method (leaving `protected` in a comment), hope for a better way in the future.

**Comment by ianthomas23** [comment]
Here is a brief explanation of what is going on here for future reference. At its simplest, imagine you have three values that are not evenly spaced, e.g. `[1, 1.2, 2]`. The histogram of these values will be `1` for the bins containing any of the three values and zero otherwise. The Cumulative Distribution Function (CDF) for this histogram is shown here:

cdf

To find the values at which to split up the palette, divide the CDF range (here `1` to `2`) into equally-spaced intervals and interpolate those boundaries using the above plot to obtain the corresponding values. I.e. draw horizontal lines at the equally-spaced CDF intervals, identify where they intersect the blue line and draw vertical lines from the intersection points to obtain the values. So we are simply interpolating the CDF line the opposite way round to how we usually think of interpolation, we have `y` values of interest and want to interpolate to get the corresponding `x` values.

This gives us the correct colors for each of our values, within some numerical tolerance governed by the number of histogram bins used.

We can also draw a colorbar to show the numerical values. These values are somewhat ambiguous as they correspond to the flat regions in the CDF plot. The code here, and the code it replaced, return values towards the top end of the flat regions (because we are interpolating from the near-vertical regions) whereas the human eye would probably prefer them more evenly spaced. This is something we could improve in the future that I am not dealing with in this PR, it needs careful consideration along with the corresponding `datashader` histogram equalisation code.

**Comment by bryevdv** [comment]
@ianthomas23 What do you think about some tests that compares concrete values? I can imagine that the behavior of `scan` could be altered while keeping the current test still passing. i.e. 
```
      for (let i = 1; i < n; i++) {
        expect(scan.binning[i]).to.be.above(sorted[i-1])
        expect(scan.binning[i]).to.be.below(sorted[i])
      }
``` 
still affords room for values to change but be unnoticed. What would be ideal is that if there are any unintentional changes that they become apparent.

**Comment by ianthomas23** [comment]
> @ianthomas23 What do you think about some tests that compares concrete values (within epsilon as appropriate)?

I've done this.

As discussed in weekly meeting I am looking into whether we need to make more significant changes to the `EqHist` code so there might be another PR for this in the near future.

**Comment by bryevdv** [comment]
Thanks @ianthomas23 !

**Pull request merged**